pkg/querier/api/consistency.go

```diff
@@ -26,81 +22,21 @@ const (
 
 var ReadConsistencies = []string{ReadConsistencyStrong, ReadConsistencyEventual}
 
-func IsValidReadConsistency(lvl string) bool {
-	return slices.Contains(ReadConsistencies, lvl)
+func IsValidReadConsistency(lvl string) error {
+	if slices.Contains(ReadConsistencies, lvl) {
+		return nil
+	}
+	return fmt.Errorf("consistency level must be one of %v", ReadConsistencies)
 }
 
-type contextKey int
-
-const consistencyContextKey contextKey = 1
-
 // ContextWithReadConsistency returns a new context with the given consistency level.
 // The consistency level can be retrieved with ReadConsistencyFromContext.
 func ContextWithReadConsistency(parent context.Context, level string) context.Context {
-	return context.WithValue(parent, consistencyContextKey, level)
+	return ContextWithAHeaderOption(parent, ReadConsistencyHeader, level)
 }
 
 // ReadConsistencyFromContext returns the consistency level from the context if set via ContextWithReadConsistency.
 // The second return value is true if the consistency level was found in the context and is valid.
 func ReadConsistencyFromContext(ctx context.Context) (string, bool) {
-	level, _ := ctx.Value(consistencyContextKey).(string)
-	return level, IsValidReadConsistency(level)
-}
-
-// ConsistencyMiddleware takes the consistency level from the X-Read-Consistency header and sets it in the context.
-// It can be retrieved with ReadConsistencyFromContext.
-func ConsistencyMiddleware() middleware.Interface {
-	return middleware.Func(func(next http.Handler) http.Handler {
-		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-			if c := r.Header.Get(ReadConsistencyHeader); IsValidReadConsistency(c) {
-				r = r.WithContext(ContextWithReadConsistency(r.Context(), c))
-			}
-			next.ServeHTTP(w, r)
-		})
-	})
-}
-
-const consistencyLevelGrpcMdKey = "__consistency_level__"
```

Comment:
if we change the metadata key, then old/new ingester/querier pairs won't be able to pass the consistency header correctly because the MD key is different.

Can you serialize each header into its own metadata key? Maybe change the implementation of `HeaderOptionsConfig` to something like this?


```go
type HeaderOptionsConfig struct {
	Headers []HeaderOptionConfig
}

type HeaderOptionConfig struct {
  HTTPHeaderName string
  GRPCMdKey      string
  Validate       func(string) error
}
```